janusgraph-core/src/main/java/org/janusgraph/diskstorage/indexing/IndexMutation.java

```diff
@@ -68,11 +75,16 @@ public void resetDelete() {
         isDeleted=false;
     }
 
-    public static final Function<IndexEntry,String> ENTRY2FIELD_FCT = new Function<IndexEntry, String>() {
+    private boolean isCollection(String field) {
+        KeyInformation keyInformation = storeRetriever.get(field);
+        return keyInformation != null && keyInformation.getCardinality() != Cardinality.SINGLE;
+    }
+
+    private final Function<IndexEntry,String> ENTRY2FIELD_FCT = new Function<IndexEntry, String>() {
         @Nullable
         @Override
         public String apply(final IndexEntry indexEntry) {
-            return indexEntry.field;
+            return isCollection(indexEntry.field) ? indexEntry.field + "." + indexEntry.value : indexEntry.field;
```

Comment:
As @chupman said here https://github.com/JanusGraph/janusgraph/pull/1547#issuecomment-493235533 this won't help if we have duplicate values (for example for `Cardinality.LIST`).

Reply:
Are you referring the very specific use case of a user with duplicate values like:
```
b = { VertexLabel : user,
  name: Robert,
  alias: [bob,bob,jimbob,bobO,bob]
}
```
and then had a transaction that added a 4th bob, but removed 2 others that those removals would get wiped out?

I would agree that this won't cover that edge case, but If you look at the test attached it will properly retain deletions if a property of LIST cardinialty is removed entirely and then readded.

Reply:
@chupman Yes, I am referring this use case. I thought about it. Maybe, if we change `consolidate` method to work with duplicates then this fix would be good. I could propose to change `consolidate` method in `Mutation` class with something like this:

```
public<V> void consolidate(Function<E,V> convertAdditions, Function<K,V> convertDeletions) {
    if (hasDeletions() && hasAdditions()) {
        Map<V, Integer> addsMap = new HashMap<>();
        for(E add : additions){
            V name = convertAdditions.apply(add);
            Integer currentSimilarAdditions = addsMap.get(name);
            if(currentSimilarAdditions == null){
                currentSimilarAdditions = 1;
            } else {
                ++currentSimilarAdditions;
            }
            addsMap.put(name, currentSimilarAdditions);
        }

        deletions.removeIf(k -> {

            V name = convertDeletions.apply(k);

            Integer currentSimilarAdditions = addsMap.get(name);

            if(currentSimilarAdditions == null){
                return false;
            }

            --currentSimilarAdditions;
            if(currentSimilarAdditions==0){
                addsMap.remove(name);
            } else {
                addsMap.put(name, currentSimilarAdditions);
            }

            return true;
        });
    }
}
```

I didn't test the above logic yet. Maybe it contains some mistakes. I just proposing to rewrite this logic so that it would be aware of duplicates. I am not yet sure, maybe it is a stupid idea. I just need some time to test it but not sure I will be able till Monday.

Reply:
I think a new, very specific issue and PR would make sense for that.

If you agree with it being a separate PR, is the only remaining comment to address making `ENTRY2FIELD_FCT` static again? Or, if there's a good reason for it not being static, having it renamed something like `entry2field_fct` so it's not confused as being static?

Reply:
`ENTRY2FIELD_FCT` should be static.
Also, I want to mention that this solution won't work if the value is an object with non overwritten `toString` method.
I agree we should create a separate issue for the discussed problem.

Reply:
The field name may contain dots, so the solution is error prone right now.
For example:
field name `hello.world`, value `!` will conflict with field name `hello`, value `world.!`.

Reply:
We could return a custom object with overwritten `equals` and `hashCode`.
The idea is to use not a string but a custom object if the field is collection. I.e.:
```
public static final Function<IndexEntry,Object> ENTRY2FIELD_FCT = new Function<IndexEntry, Object>() {
    @Nullable
    @Override
    public Object apply(final IndexEntry indexEntry) {

        if(isCollection(indexEntry.field)){
            return new FieldDefinition(indexEntry.field, indexEntry.value);
        }

        return indexEntry.field;
    }
};
```

FieldDefinition could be something like:
```
class FieldDefinition {
    private String field;
    private Object value;

    public FieldDefinition(String field, Object value) {
        this.field = field;
        this.value = value;
    }

    @Override
    public boolean equals(Object o){
        if(o instanceof FieldDefinition){
            return Objects.equals(field, ((FieldDefinition) o).field) && 
                Objects.equals(value, ((FieldDefinition) o).value);
        }
        return false;
    }
    
    @Override
    public int hashCode(){
        return Objects.hash(field, value);
    }
}
```

In such situation we won't have above problems with field name and custom objects don't need to overwrite `toString` method.

Reply:
@porunov I think this would work if `ENTRY2FIELD_FCT` returned the `hashCode()` instead of the FieldDefinition  object. I might be missing it, but I don't think we even need the equals method since the check is `contains`, not `equal`. Nice write-up! I was still thinking about this and really like your approach.

Separately how do you want to handle this PR? Right now my thoughts are `ENTRY2FIELD_FCT ` needs to be made static and it needs to be rebased onto 0.3. It doesn't create any new issues. It just doesn't completely fix every use case. I also lays good groundwork to fully fix the problem. My suggestion would be to finish cleaning this PR up and do the other fixes in a separate follow-up PR. 

This bug and underlying concept is at least moderately complicated and I think it's better to go with small focused commits as much as possible. That being said it's totally your call.

Reply:
> I think this would work if ENTRY2FIELD_FCT returned the hashCode() instead of the FieldDefinition object. I might be missing it, but I don't think we even need the equals method since the check is contains, not equal. Nice write-up! I was still thinking about this and really like your approach.

It isn't quite right to use `hashCode` only because `hashCode` is used in hash table (aka HashSet) only for bucket identification. If the collision acquires then `equals` method is executed to add or overwrite the object. Notice, `hashCode` is `int`, that is why it is easy to get collisions with `long` data type. So, we have to overwrite both `hashCode` and `equals` methods. Otherwise our `Set` will not work properly for `add` operation.

> Separately how do you want to handle this PR? Right now my thoughts are ENTRY2FIELD_FCT needs to be made static and it needs to be rebased onto 0.3.

I would suggest do next things:
1) Use `FieldDefinition` and rewrite `ENTRY2FIELD_FCT` as I shown above.
2) Make `ENTRY2FIELD_FCT` static.
3) Rebase against `0.3` branch.

I am open for suggestions and criticism of this solutions.

Reply:
https://github.com/JanusGraph/janusgraph/blob/ff4fe7148a5197eed65bf0b90f2b86c480522f51/janusgraph-core/src/main/java/org/janusgraph/diskstorage/Mutation.java#L151-L157

From what I understand the `ENTRY2FIELD_FCT` function being passed to `consolidate` doesn't affect that values in the `additions` and `deletions` Lists for the mutation. When consolidate is called right before a commit `adds` is populated based of logic in `ENTRY2FIELD_FCT`. `adds` is being populated with `IndexEntry.field`, which is a string, not the value so there should not be any collision concerns. 

I like hashing better than key, value concatenation because on large strings we'd be eating up memory on creating huge strings that won't be interned for comparison.

Reply:
> From what I understand the ENTRY2FIELD_FCT function being passed to consolidate doesn't affect that values in the additions and deletions Lists for the mutation. When consolidate is called right before a commit adds is populated based of logic in ENTRY2FIELD_FCT. adds is being populated with IndexEntry.field, which is a string, not the value so there should not be any collision concerns.

`ENTRY2FIELD_FCT` function decides what objects will be included in `adds` HashSet. Then it removes similar adds from `deletions`. Right now all objects in `adds` are `String` with `IndexEntry.field`.
If we want to fix problems for `Cardinality.SET` we should include in `adds` not only objects that are represented by a specific `field` but also `value`.
As we know, the value is `Object` and it can be anything. My proposal is to return not a `String` in `ENTRY2FIELD_FCT` but an `Object` which represents both `field` and `value`.
I am not sure if we are talking about same collision.
Do you propose to use something like?:
```
public static final Function<IndexEntry,Object> ENTRY2FIELD_FCT = new Function<IndexEntry, Object>() {
    @Nullable
    @Override
    public Object apply(final IndexEntry indexEntry) {

        if(isCollection(indexEntry.field)){
            return Objects.hash(indexEntry.field, indexEntry.value);
        }

        return indexEntry.field;
    }
};
```

If yes, then it is a wrong solution because of hashCode collision. 
You said that if we have a unique `indexEntry.field` then we won't get a collision. It is wrong. We can get a collision here. Here are facts:
- `int` is `32` bits. `indexEntry.field` can be more then `32` bits. (Can get collision based only on this fact)
- `value` is anything. For example, value can be `long` which is `64` bits. (Can get collision based only on this fact)
- unique value doesn't guarantee a unique hashCode. (Can get collision based only on this fact)

We are calculating one `hashCode` for `indexEntry.field` and `indexEntry.value`. That is why the probability of collision even higher.

We must use a custom object like `FieldDefinition` (shown here https://github.com/JanusGraph/janusgraph/pull/1575#discussion_r286517381) with overwritten `hashCode` **and** `equals` to be added into `adds`. In such situation if the collision acquires when we calculate `hashCode` then HashSet will resolve this collision automatically by executing `equals` method. In your situation you would pass `Integer` objects of calculated hash which would be overwritten in a case of collision.

Based on these facts I am proposing the solution described here: https://github.com/JanusGraph/janusgraph/pull/1575#discussion_r286517381

I am still open for discussion. I could miss something also.

Reply:
So 3 things I suppose:
1. I was narrowly (and incorrectly) focused on a collision in respect to the aforementioned issue where a `key` containing a `.` could collide with a different `key + . + value`.
2. After brushing up on the contract between equals() and hashCode() I agree with your approach.and reasoning.
3. Thanks for the extra explanation. I'll look at this again after I do some more reading.